hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java

```diff
@@ -304,9 +305,32 @@ private void addResolvedEntityInsertAction(AbstractEntityInsertAction insert) {
 
 	@SuppressWarnings("unchecked")
 	private <T extends Executable & Comparable & Serializable> void addAction(Class<T> executableClass, T action) {
+		if ( action instanceof EntityUpdateAction && trySquashEntityUpdateAction( (EntityUpdateAction) action ) ) {
+			return;
+		}
 		EXECUTABLE_LISTS_MAP.get( executableClass ).getOrInit( this ).add( action );
 	}
 
+	private boolean trySquashEntityUpdateAction(EntityUpdateAction action) {
+		// merge update into insert if possible
+		if ( insertions != null ) {
+			Iterator<AbstractEntityInsertAction> it = insertions.iterator();
+			while ( it.hasNext() ) {
+				AbstractEntityInsertAction insertAction = it.next();
+				if ( insertAction.getInstance() == action.getInstance() ) {
+					Object[] oldState = insertAction.getState();
+					Object[] newState = action.getState();
+					int[] dirtyFields = action.getDirtyFields();
+					for (int i = 0; i < dirtyFields.length; i++) {
+						oldState[dirtyFields[i]] = newState[dirtyFields[i]];
+					}
+					return true;
+				}
+			}
```

Comment:
Hibernate codebase has the tradition to use `final` keyword aggressively for temporary variables within method. Would be great to adopt it for consistency reason.